applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.h

```diff
@@ -0,0 +1,449 @@
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    /**
+    * @brief Get on rVariable a Vector Value from the Element Constitutive Law
+    * @param rVariable The variable we want to get
+    * @param rValues The results in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void GetValueOnIntegrationPoints(
+        const Variable<Vector>& rVariable,
+        std::vector<Vector>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    /**
+    * @brief Get on rVariable a Matrix Value from the Element Constitutive Law
+    * @param rVariable The variable we want to get
+    * @param rValues The results in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void GetValueOnIntegrationPoints(
+        const Variable<Matrix>& rVariable,
+        std::vector<Matrix>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        std::stringstream buffer;
+        buffer << "\"BaseDiscreteElement\" #" << Id();
+        return buffer.str();
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << "\"BaseDiscreteElement\" #" << Id();
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const {
+        pGetGeometry()->PrintData(rOStream);
+    }
+    ///@}
+
+protected:
+
+    ///@name Protected member Variables
+    ///@{
+    std::vector<ConstitutiveLaw::Pointer> mConstitutiveLawVector; /// The vector containing the constitutive laws
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    /**
+    * @brief It initializes the material
+    */
+    virtual void InitializeMaterial();
+
+    /**
+    * This functions calculates both the RHS and the LHS
+    * @param rLeftHandSideMatrix: The LHS
+    * @param rRightHandSideVector: The RHS
+    * @param rCurrentProcessInfo: The current process info instance
+    * @param CalculateStiffnessMatrixFlag: The flag to set if compute the LHS
+    * @param CalculateResidualVectorFlag: The flag to set if compute the RHS
+    */
+    virtual void CalculateAll(
+        MatrixType& rLeftHandSideMatrix,
+        VectorType& rRightHandSideVector,
+        ProcessInfo& rCurrentProcessInfo,
+        const bool CalculateStiffnessMatrixFlag,
+        const bool CalculateResidualVectorFlag);
+
+    /**
+    * Jacobian computes the mapping from Geometry to Parameter Space.
+    *
+    * @param[in] DN_De derivatives of shape functions in two directions.
+    * @param[out] Jacobian calculated Jacobian. 
+    * @param[in] rWorkingSpaceDimension GeometrySpace coordinates. For surfaces default 3.
+    * @param[in] rLocalSpaceDimension ParameterSpace coordinates For surfaces default 2.
+    */
+    void Jacobian(const Matrix& DN_De,
```

Comment:
didn't you say this should be virtual?

Reply:
const int& == const int

Reply:
The function is valid for any geometry object due to the space dimension inputs.

Reply:
I changed the int& to int.